src/rust/bitbox02-rust/src/async_usb.rs

```diff
@@ -0,0 +1,175 @@
+// Copyright 2020 Shift Cryptosecurity AG
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! This module provides the executor for tasks that are spawned with an API request and deliver a
+//! USB response.
+
+extern crate alloc;
+
+use crate::bb02_async::{spin as spin_task, Task};
+use alloc::boxed::Box;
+use alloc::vec::Vec;
+use core::cell::RefCell;
+use core::task::Poll;
+
+type UsbOut = Vec<u8>;
+type UsbIn = Vec<u8>;
+
+/// Describes the global state of an api query. The documentation of
+/// the variants apply to the HWW stack, but have analogous meaning in
+/// the U2F stack.
+enum UsbTaskState {
+    /// Waiting for a new query, nothing to do.
+    Nothing,
+    /// A query came in which launched a task, which is now running (e.g. user is entering a
+    /// password).
+    Running(Task<UsbOut>),
+    /// The task has finished and written the result, so the USB response is available. We are now
+    /// waiting for the client to fetch it (HWW_REQ_RETRY). For short-circuited or non-async api
+    /// calls, the result might be returned immediately in response to HWW_REQ_NEW.
+    ResultAvailable(UsbOut),
+}
+
+/// Executor main state. Currently we only have at most one task at a time (usb api processing
+/// task).
+///
+/// It is `mut` because without it, Rust requires thread safety via
```

Comment:
Whats the drawback to `unsafe impl Send for RefCell<UsbTaskState>`? Should be safe because we are single threaded.

Reply:
It seems to be both prefixing `mut` as well as marking with with `Sync` (this is enough for immutable) are workarounds, not sure which is better.

Sidenote: can't impl foreign impl on foreign trait; it would work if I wrapped it.

`Sync` marks it as thread safe, which it is not. Of course we are single threaded, but it feels bad to mark it as thread safe. 

The whole thing actually also works without a RefCell:

`static mut USB_TASK_STATE: UsbTaskState = UsbTaskState::Nothing;`

The RefCell only is a small dev help so that it's not borrowed mutably twice (e.g. by accidentally calling spin() inside of spawn()  etc.), but in spirit the var is mutable, so I didn't mind the `mut` in front.

Maybe `Sync` is a safer workaround, because then the global cannot actually be mutated (sidestepping RefCell)?

Reply:
Pushed a commit to use unsafe `Sync`, it does look better -> it reduces the `unsafe` into one place and removes the danger of mutating the state without RefCell.